src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1008,83 +1008,31 @@ int32_t SystemNative_PosixFAdvise(intptr_t fd, int64_t offset, int64_t length, i
 #endif
 }
 
-int32_t SystemNative_PosixFAllocate(intptr_t fd, int64_t offset, int64_t length)
+int32_t SystemNative_FAllocate(intptr_t fd, int64_t offset, int64_t length)
 {
     assert_msg(offset == 0, "Invalid offset value", (int)offset);
 
     int fileDescriptor = ToFileDescriptor(fd);
     int32_t result;
-#if HAVE_POSIX_FALLOCATE64 // 64-bit Linux
-    while ((result = posix_fallocate64(fileDescriptor, (off64_t)offset, (off64_t)length)) == EINTR);
-#elif HAVE_POSIX_FALLOCATE // 32-bit Linux
-    while ((result = posix_fallocate(fileDescriptor, (off_t)offset, (off_t)length)) == EINTR);
+#if HAVE_FALLOCATE // Linux
+    while ((result = fallocate(fileDescriptor, FALLOC_FL_KEEP_SIZE, (off_t)offset, (off_t)length)) == EINTR);
```

Comment:
On Linux, no.
On macOS, I think so.

Another change I made in this PR is for `preallocationSize` to require `FileMode.Create*` or `FileMode.Truncate*`. So we'll be calling this on an empty file (and consequently can't be truncating).

Reply:
By truncate, I meant overflow, i.e. might off_t be 32-bit on some systems, in which case this might do the wrong thing?

Reply:
> might off_t be 32-bit on some systems

We're compiling `-D_FILE_OFFSET_BITS=64`, so these will be 64-bit.

> Wasn't sure in your response how changing behavior for FileMode would impact that.

I misread your question. The `fcntl` can cause the file itself to get truncated on macOS. But that's not an issue here since we're calling this function with an empty file.

Reply:
> We're compiling -D_FILE_OFFSET_BITS=64, so these will be 64-bit.

Interesting.  While we used to do so in some circumstances (I don't think we always did it), it looks like like as of a few weeks ago we now do it everywhere:
https://github.com/dotnet/runtime/pull/58096/files#diff-36ef498ca978a03472b1216f6e3c603e6e3950d27bc7f70972c078b444f5e8b9R280-R281